feat(core): Add metric for active workflow count

juusotn8n opened the pull request:
## Summary

Adds a new metric for active workflow count. This metric is checked every time metrics are collected. Due to the workflow count being potentially quite high, we also cache the result and utilize the cached value.

This metric needs to be enabled separately and the TTL can be configured.

## Related Linear tickets, Github issues, and Community forum posts

https://linear.app/n8n/issue/CP-1797/add-metric-for-active-workflow-count

<!--
Include links to **Linear ticket** or Github issue or Community forum post.
Important in order to close *automatically* and provide context to reviewers.
-->
<!-- Use "closes #<issue-number>", "fixes #<issue-number>", or "resolves #<issue-number>" to automatically close issues when the PR is merged. -->


## Review / Merge checklist

- [x] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) <!--
   **Remember, the title automatically goes into the changelog.
   Use `(no-changelog)` otherwise.**
-->
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [x] Tests included. <!--
   A bug is not considered fixed, unless a test is added to prevent it from happening again.
   A feature is not complete without tests.
-->
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

juusotn8n committed:
feat(core): add metric for active workflow count

Adds a new metric for active workflow count. This metric is checked
every time metrics are collected. Due to the workflow count being
potentially quite high, we also cache the result and utilize the
cached value.

This metric needs to be enabled separately and the TTL can be
configured.

ivov commented on packages/cli/src/databases/repositories/workflow.repository.ts:
```diff
@@ -57,6 +57,12 @@ export class WorkflowRepository extends Repository<WorkflowEntity> {
 		return activeWorkflows.map((workflow) => workflow.id);
 	}
 
+	async getActiveCount() {
+		return await this.count({
+			where: { active: true },
+		});
+	}
```
Curious why not get this from memory instead? See `ActiveWorkflows.ts`

juusotn8n replied:
I wasn't aware this was kept up to date somewhere in memory. I can try it out, would make the implementation simpler.

ivov replied:
Turns out webhook-based active workflows are excluded from `ActiveWorkflows`

tomi commented on packages/@n8n/config/src/configs/endpoints.config.ts:
```diff
@@ -57,6 +57,14 @@ class PrometheusMetricsConfig {
 	queueMetricsInterval: number = 20;
+
+	/** Whether to include metric for active workflow count */
+	@Env('N8N_METRICS_INCLUDE_ACTIVE_WORKFLOW_COUNT_METRIC')
+	includeActiveWorkflowCountMetric: boolean = false;
+
+	/** How often (in seconds) to update active workflow metric */
+	@Env('N8N_METRICS_ACTIVE_WORKFLOW_METRIC_INTERVAL')
+	activeWorkflowCountInterval: number = 60;
```
Do we really need yet more configs? Or could we simply add them for all? Asking because I would much rather see us reduce the amount of things one can configure to simplify things

juusotn8n replied:
The metric itself is configurable because most of the other metrics are, I don't see why it couldn't be activated just baseline if metrics are enabled. As far as the cache TTL goes, I think 60 seconds is a sensible default but whether or not that is useful depends on how often you ping `/metrics`. So I'm not sure I'd make that just a static value

tomi reviewed (dismissed):
LGTM 🚀 Very nice tests 👌

juusotn8n committed:
chore: use pre-existing constant for milliseconds conversion

juusotn8n committed:
feat: remove toggle for active workflow count metric

Include metric if metrics are enabled. Leave cache TTL as
configurable, since a good time to cache might depend on how
often metrics is pinged

tomi reviewed (approved):
🚀 

P.S. Instead of rebasing please merge `master` into the branch, as rebasing unfortunately breaks Github's show "Changes since your last review" 🙏 

<img width="410" alt="image" src="https://github.com/user-attachments/assets/3b635e7f-3f87-4f3c-a2fc-25c831d41064" />

juusotn8n committed:
Merge branch 'master' into cp-1797-add-metric-for-active-workflow-count

janober commented:
Got released with [`n8n@1.82.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.82.0)
